src/integrators/ptracer.cpp

```diff
@@ -0,0 +1,245 @@
+#include <random>
+
+#include <enoki/morton.h>
+#include <mitsuba/core/fwd.h>
+#include <mitsuba/core/plugin.h>
+#include <mitsuba/core/progress.h>
+#include <mitsuba/core/properties.h>
+#include <mitsuba/core/ray.h>
+#include <mitsuba/render/bsdf.h>
+#include <mitsuba/render/emitter.h>
+#include <mitsuba/render/integrator.h>
+#include <mitsuba/render/records.h>
+#include <mitsuba/render/sampler.h>
+
+#include <fstream>
+
+// #define MTS_DEBUG_PTRACER_PATHS "/tmp/ptracer.obj"
+#if defined(MTS_DEBUG_PTRACER_PATHS)
+#include <fstream>
+namespace {
+static size_t export_counter = 0;
+} // namespace
+#endif
+
+NAMESPACE_BEGIN(mitsuba)
+
+/**
+ * Traces rays from the light source and attempts to connect them to the sensor
+ * at each bounce.
+ */
+template <typename Float, typename Spectrum>
+class ParticleTracerIntegrator : public LightTracerIntegrator<Float, Spectrum> {
```

Comment:
The particle tracer is a sufficiently special snowflake that I don't think we will have another integrator like it. Given that, I think that it would be better to have the specialization of the `render()` method directly here in the class.